File: wlt/persist.py
"""Persist the nftables src2mark map to a reloadable include file."""

import ipaddress
import json
import logging
import os
import signal
import subprocess
import tempfile
import threading
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from .config import NftablesConfig, load_config

logger = logging.getLogger(__name__)

SNAPSHOT_PATH = Path("/etc/nftables/wlt_src2mark.conf")
SAVE_INTERVAL_SECONDS = 300


@dataclass(frozen=True)
class PersistedEntry:
    ip: str
    mark: int
    timeout: int | None = None
    expires: int | None = None


def parse_map_entries(data: dict[str, Any]) -> list[PersistedEntry]:
    map_data = next(
        (
            item["map"]
            for item in data.get("nftables", [])
            if isinstance(item, dict) and isinstance(item.get("map"), dict)
        ),
        None,
    )
    if map_data is None:
        raise ValueError("nftables JSON does not contain a map")

    entries: list[PersistedEntry] = []

    for item in map_data.get("elem", []):
        if not isinstance(item, list) or len(item) != 2:
            raise ValueError(f"Unexpected nftables map element: {item!r}")

        key, raw_mark = item
        timeout = None
        expires = None
        if isinstance(key, dict) and isinstance(key.get("elem"), dict):
            element = key["elem"]
            raw_ip = element.get("val")
            timeout = _optional_seconds(element.get("timeout"))
            expires = _optional_seconds(element.get("expires"))
        else:
            raw_ip = key

        ip = str(ipaddress.ip_address(str(raw_ip)))
        mark = raw_mark if isinstance(raw_mark, int) else int(str(raw_mark), 0)
        if mark < 0 or mark > 0xFFFFFFFF:
            raise ValueError(f"Invalid mark for {ip}: {mark}")

        entries.append(
            PersistedEntry(
                ip=ip,
                mark=mark,
                timeout=timeout,
                expires=expires,
            )
        )

    return sorted(entries, key=lambda entry: ipaddress.ip_address(entry.ip))


def _optional_seconds(value: Any) -> int | None:
    if value is None:
        return None
    seconds = int(value)
    return max(seconds, 1)


def render_snapshot(
    config: NftablesConfig, sections: list[tuple[str, list[PersistedEntry]]]
) -> str:
    lines = [
        "# Managed by wlt-persist. Manual changes will be overwritten.",
        "# Timeout counters resume from the saved remaining time after reload.",
    ]
    for map_name, entries in sections:
        if not entries:
            continue
        lines.append(f"add element {config.family} {config.table} {map_name} {{")
        for index, entry in enumerate(entries):
            options = []
            if entry.timeout is not None:
                options.append(f"timeout {entry.timeout}s")
            if entry.expires is not None:
                options.append(f"expires {entry.expires}s")
            option_text = f" {' '.join(options)}" if options else ""
            comma = "," if index < len(entries) - 1 else ""
            lines.append(f"    {entry.ip}{option_text} : {hex(entry.mark)}{comma}")
        lines.append("}")
    return "\n".join(lines) + "\n"


def map_names(config: NftablesConfig) -> list[str]:
    names = [config.map]
    if config.map_v6 and config.map_v6 not in names:
        names.append(config.map_v6)
    return names


def fetch_map(config: NftablesConfig, map_name: str) -> dict[str, Any]:
    result = subprocess.run(
        [
            "nft",
            "--json",
            "list",
            "map",
            config.family,
            config.table,
            map_name,
        ],
        capture_output=True,
        text=True,
        check=True,
    )
    return json.loads(result.stdout)


def write_if_changed(path: Path, content: str) -> bool:
    encoded = content.encode()
    try:
        if path.read_bytes() == encoded:
            return False
    except FileNotFoundError:
        pass

    path.parent.mkdir(parents=True, exist_ok=True)
    fd, temp_name = tempfile.mkstemp(prefix=f".{path.name}.", dir=path.parent)
    temp_path = Path(temp_name)
    try:
        os.fchmod(fd, 0o644)
        with os.fdopen(fd, "wb") as f:
            f.write(encoded)
            f.flush()
            os.fsync(f.fileno())
        os.replace(temp_path, path)
        directory_fd = os.open(path.parent, os.O_RDONLY | os.O_DIRECTORY)
        try:
            os.fsync(directory_fd)
        finally:
            os.close(directory_fd)
    finally:
        temp_path.unlink(missing_ok=True)
    return True


def save_snapshot(config: NftablesConfig, path: Path = SNAPSHOT_PATH) -> bool:
    sections: list[tuple[str, list[PersistedEntry]]] = []
    total = 0
    for map_name in map_names(config):
        entries = parse_map_entries(fetch_map(config, map_name))
        sections.append((map_name, entries))
        total += len(entries)
    changed = write_if_changed(path, render_snapshot(config, sections))
    if changed:
        logger.info("Saved %d src2mark entries to %s", total, path)
    else:
        logger.debug("src2mark snapshot is unchanged")
    return changed


def _save_safely(config: NftablesConfig, path: Path) -> None:
    try:
        save_snapshot(config, path)
    except (
        json.JSONDecodeError,
        OSError,
        subprocess.SubprocessError,
        ValueError,
    ) as e:
        logger.error("Failed to save src2mark snapshot: %s", e)


def run(
    config: NftablesConfig,
    path: Path = SNAPSHOT_PATH,
    interval: int = SAVE_INTERVAL_SECONDS,
) -> None:
    stop_event = threading.Event()

    def request_stop(signum, frame):
        stop_event.set()

    signal.signal(signal.SIGTERM, request_stop)
    signal.signal(signal.SIGINT, request_stop)

    _save_safely(config, path)
    while not stop_event.wait(interval):
        _save_safely(config, path)
    _save_safely(config, path)


def main() -> None:
    logging.basicConfig(level=logging.INFO)
    run(load_config().nftables)


if __name__ == "__main__":
    main()

File: src/persist.rs
//! Persist the nftables src2mark maps to a reloadable include file.

use std::net::IpAddr;
use std::path::Path;
use std::time::Duration;

use anyhow::{Context, Result};
use tokio_util::sync::CancellationToken;

use crate::config::{AppConfig, PersistConfig};
use crate::nft::{Nft, NftEntry};

fn map_names(cfg: &AppConfig) -> Vec<&str> {
    let mut names = vec![cfg.nftables.map.as_str()];
    if let Some(map_v6) = cfg.nftables.map_v6.as_deref()
        && !names.contains(&map_v6)
    {
        names.push(map_v6);
    }
    names
}

/// nftables reports timeout/expires in whole seconds; clamp to at least 1s so
/// the rendered element stays valid.
fn optional_seconds(value: Option<u32>) -> Option<u32> {
    value.map(|seconds| seconds.max(1))
}

fn sort_entries(mut entries: Vec<NftEntry>) -> Vec<NftEntry> {
    entries.sort_by(|a, b| {
        let (ka, kb) = (a.ip.parse::<IpAddr>().ok(), b.ip.parse::<IpAddr>().ok());
        ka.cmp(&kb).then_with(|| a.ip.cmp(&b.ip))
    });
    entries
}

pub fn render_snapshot(family: &str, table: &str, sections: &[(&str, Vec<NftEntry>)]) -> String {
    let mut lines = vec![
        "# Managed by wlt-persist. Manual changes will be overwritten.".to_owned(),
        "# Timeout counters resume from the saved remaining time after reload.".to_owned(),
    ];
    for (map_name, entries) in sections {
        if entries.is_empty() {
            continue;
        }
        lines.push(format!("add element {family} {table} {map_name} {{"));
        for (index, entry) in entries.iter().enumerate() {
            let mut options = String::new();
            if let Some(timeout) = optional_seconds(entry.timeout) {
                options.push_str(&format!(" timeout {timeout}s"));
            }
            if let Some(expires) = optional_seconds(entry.expires) {
                options.push_str(&format!(" expires {expires}s"));
            }
            let comma = if index + 1 < entries.len() { "," } else { "" };
            lines.push(format!(
                "    {}{} : {:#x}{}",
                entry.ip, options, entry.mark, comma
            ));
        }
        lines.push("}".to_owned());
    }
    lines.join("\n") + "\n"
}

/// Atomically replace `path` with `content` if it differs; fsyncs both the
/// file and its directory. Returns whether a write happened.
pub fn write_if_changed(path: &Path, content: &str) -> Result<bool> {
    if std::fs::read(path).is_ok_and(|existing| existing == content.as_bytes()) {
        return Ok(false);
    }

    let dir = path
        .parent()
        .context("snapshot path has no parent directory")?;
    std::fs::create_dir_all(dir)?;
    let file_name = path
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("snapshot");
    let mut temp = tempfile::Builder::new()
        .prefix(&format!(".{file_name}."))
        .tempfile_in(dir)?;
    {
        use std::io::Write;
        use std::os::unix::fs::PermissionsExt;
        temp.as_file()
            .set_permissions(std::fs::Permissions::from_mode(0o644))?;
        temp.write_all(content.as_bytes())?;
        temp.as_file().sync_all()?;
    }
    temp.persist(path).context("failed to replace snapshot")?;
    std::fs::File::open(dir)?.sync_all()?;
    Ok(true)
}

async fn save_snapshot(cfg: &AppConfig, nft: &Nft, path: &Path) -> Result<bool> {
    let mut sections = Vec::new();
    let mut total = 0;
    for map_name in map_names(cfg) {
        let entries = sort_entries(nft.list_entries(map_name).await?);
        total += entries.len();
        sections.push((map_name, entries));
    }
    let content = render_snapshot(nft.family_str(), nft.table(), &sections);
    let changed = write_if_changed(path, &content)?;
    if changed {
        tracing::info!("saved {total} src2mark entries to {}", path.display());
    } else {
        tracing::debug!("src2mark snapshot is unchanged");
    }
    Ok(changed)
}

async fn save_safely(cfg: &AppConfig, nft: &Nft, path: &Path) {
    if let Err(e) = save_snapshot(cfg, nft, path).await {
        tracing::error!("failed to save src2mark snapshot: {e:#}");
    }
}

pub async fn run(
    cfg: std::sync::Arc<AppConfig>,
    nft: Nft,
    persist: PersistConfig,
    shutdown: CancellationToken,
) -> Result<()> {
    let interval = Duration::from_secs(persist.interval);
    save_safely(&cfg, &nft, &persist.path).await;
    loop {
        tokio::select! {
            _ = shutdown.cancelled() => break,
            _ = tokio::time::sleep(interval) => save_safely(&cfg, &nft, &persist.path).await,
        }
    }
    // Final snapshot so a clean shutdown loses nothing.
    save_safely(&cfg, &nft, &persist.path).await;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn entry(ip: &str, mark: u32, timeout: Option<u32>, expires: Option<u32>) -> NftEntry {
        NftEntry {
            ip: ip.into(),
            mark,
            timeout,
            expires,
        }
    }

    #[test]
    fn renders_permanent_and_timed_entries() {
        let sections = vec![(
            "src2mark",
            vec![
                entry("10.0.0.1", 0x1, None, None),
                entry("10.0.0.2", 0x1200, Some(3600), Some(0)),
            ],
        )];
        let text = render_snapshot("inet", "wlt", &sections);
        assert_eq!(
            text,
            "# Managed by wlt-persist. Manual changes will be overwritten.\n\
             # Timeout counters resume from the saved remaining time after reload.\n\
             add element inet wlt src2mark {\n\
             \x20   10.0.0.1 : 0x1,\n\
             \x20   10.0.0.2 timeout 3600s expires 1s : 0x1200\n\
             }\n"
        );
    }

    #[test]
    fn skips_empty_sections() {
        let sections = vec![
            ("src2mark", vec![]),
            ("src2mark6", vec![entry("fd00::1", 0x30, None, None)]),
        ];
        let text = render_snapshot("inet", "wlt", &sections);
        assert!(!text.contains("add element inet wlt src2mark {"));
        assert!(text.contains("add element inet wlt src2mark6 {\n    fd00::1 : 0x30\n}"));
    }

    #[test]
    fn sorts_entries_by_ip() {
        let entries = vec![
            entry("10.0.0.10", 1, None, None),
            entry("10.0.0.2", 2, None, None),
            entry("9.9.9.9", 3, None, None),
        ];
        let sorted = sort_entries(entries);
        let ips: Vec<&str> = sorted.iter().map(|e| e.ip.as_str()).collect();
        assert_eq!(ips, ["9.9.9.9", "10.0.0.2", "10.0.0.10"]);
    }

    #[test]
    fn write_if_changed_only_writes_on_change() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("snapshot.conf");
        assert!(write_if_changed(&path, "hello\n").unwrap());
        assert!(!write_if_changed(&path, "hello\n").unwrap());
        assert!(write_if_changed(&path, "world\n").unwrap());
        assert_eq!(std::fs::read_to_string(&path).unwrap(), "world\n");
    }
}
